blog/_src/posts/2020-10-15-shallow-mailing-list.md

```diff
@@ -0,0 +1,414 @@
+That second contract raises the error message.
+
+
+### How's transient?
+
+**Shallow** runs successfully.
+The cast looks for a hash, does not make a contract, and ignores the inferred
+ type going forward.
+
+- - -
+
+
+## Same-Arity Functions in a Case Lambda
+
+Original message : <https://groups.google.com/g/racket-users/c/BDrrgW0axGQ/m/P31NxeGHAAAJ>
+
+
+#### On 2019-07-05, Ryan Kramer wrote:
+
+> In the code below, can `maybe-car` have the given type [....]?
+
+```
+#lang typed/racket
+
+(module untyped racket
+  (provide maybe-car)
+  (define (maybe-car x)
+    (cond
+      [(pair? x) (car x)]
+      [else x])))
+
+(require/typed
+ 'untyped
+ [maybe-car (All (a b) (case->
+                        (-> (Pairof a b) a)
+                        (-> a a)))])
+```
+
+> [Current error:]
+
+```
+Type Checker:
+ Type (All (a b) (case-> (-> (Pairof a b) a) (-> a a)))
+  could not be converted to a contract:
+   function type has two cases of arity 1
+```
+
+### What's going on?
+
+**Deep** tries to enforce the type with a Racket `or/c` contract, but or/c
+ does not allow two function contracts with the same arity because has to
+ pick one to apply at run-time, but cannot.
+
+Note, the "Type Checker" error message is much better than what or/c would
+ give on its own.
+
+
+### How's transient?
+
+**Shallow** simply checks that maybe-car accepts both arities inside the
+ case-> type.
+The code runs without error.
+Later, when the function gets applied in typed code, Shallow spot-checks the
+ results.
+
+- - -
+
+
+### Immutable Type Affects Untyped Code
+
+Original message : <https://groups.google.com/g/racket-users/c/UD20HadJ9Ec/m/Lmuw0U8mBwAJ>
+
+#### On 2020-02-17, Bertrand Augereau wrote:
+
+> Hello everybody,
+> I'm trying to gradually type my script to make it a proper app (yes
+> I'm a static-ish guy) and I have an issue (Racket 7.6 CS).
+
+```
+; racket_mod.rkt:
+#lang racket
+
+(provide (struct-out s))
+(provide list-of-s)
+(provide set-list-of-s!)
+
+(struct s (a))
+(define list-of-s '())
+(define (set-list-of-s! los)
+  (set! list-of-s los))
+```
+
+```
+; racket_mod_typed.rkt:
+#lang typed/racket
+
+(provide (struct-out s2))
+(provide list-of-s2)
+(provide set-list-of-s2!)
+
+(struct s2 ([a : Natural]))
+(define list-of-s2 '())
+(define (set-list-of-s2! [los : (Listof s2)])
+  (set! list-of-s2 los))
+```
+
+```
+; racket_main.rkt:
+#lang racket
+
+(require "racket_mod.rkt")
+(require "racket_mod_typed.rkt")
+
+(define los (list (s 1) (s 2)))
+(set-list-of-s! los)
+(displayln list-of-s)
+
+(define los2 (list (s2 1) (s2 2)))
+(set-list-of-s2! los2)
+(displayln list-of-s2)
+```
+
+> list-of-s2 is empty and list-of-s is not, the only difference seems to
+> be the type annotations.
+> Can someone help me ? :)
+
+
+### What's going on?
+
+**Deep** enforces the type of `list-of-s2` with a listof contract, which
+ ends up making a copy of the original (empty) list as it traverses and
```

Comment:
I only partially got the whole explanation. Maybe some references giving it a temporal angle of view could help, e.g. at which point the copy created. Maybe something else.